xds/src/main/java/io/grpc/xds/LeastRequestLoadBalancer.java

```diff
@@ -0,0 +1,417 @@
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.CONNECTING;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.READY;
+import static io.grpc.ConnectivityState.SHUTDOWN;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+import static io.grpc.xds.LeastRequestLoadBalancerProvider.DEFAULT_CHOICE_COUNT;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import io.grpc.Attributes;
+import io.grpc.ClientStreamTracer;
+import io.grpc.ClientStreamTracer.StreamInfo;
+import io.grpc.ConnectivityState;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.LoadBalancer;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.xds.ThreadSafeRandom.ThreadSafeRandomImpl;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.Nonnull;
+
+/**
+ * A {@link LoadBalancer} that provides least request load balancing based on
+ * outstanding request counters.
+ * It works by sampling a number of subchannels and picking the one with the
+ * fewest amount of outstanding requests.
+ * The default sampling amount of two is also known as
+ * the "power of two choices" (P2C).
+ */
+final class LeastRequestLoadBalancer extends LoadBalancer {
+  @VisibleForTesting
+  static final Attributes.Key<Ref<ConnectivityStateInfo>> STATE_INFO =
+      Attributes.Key.create("state-info");
+  @VisibleForTesting
+  static final Attributes.Key<AtomicInteger> IN_FLIGHTS =
+      Attributes.Key.create("in-flights");
+
+  private final Helper helper;
+  private final ThreadSafeRandom random;
+  private final Map<EquivalentAddressGroup, Subchannel> subchannels =
+      new HashMap<>();
+
+  private ConnectivityState currentState;
+  private LeastRequestPicker currentPicker = new EmptyPicker(EMPTY_OK);
+  private int choiceCount = DEFAULT_CHOICE_COUNT;
+
+  LeastRequestLoadBalancer(Helper helper) {
+    this(helper, ThreadSafeRandomImpl.instance);
+  }
+
+  LeastRequestLoadBalancer(Helper helper, ThreadSafeRandom random) {
+    this.helper = checkNotNull(helper, "helper");
+    this.random = checkNotNull(random, "random");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    LeastRequestConfig config =
+        (LeastRequestConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+    // Config may be null if least_request is used outside xDS
```

Comment:
Is that so? I guess it may be if it becomes the default LB? We may want to fix that in gRPC.

Reply:
When trying this LB implementation together with a custom dns-based resolver it ended up being null here.
I haven't dug into exactly why this is the case however the custom resolver never sets any service config in the resolution result and also `loadBalancingPolicyConfig` is marked with `@Nullable` [here](https://github.com/grpc/grpc-java/blob/80f1be03f0dc347125765bd9360848f221a7f174/api/src/main/java/io/grpc/LoadBalancer.java#L174-L175) which also is a reason to do a null check.